component/grant-type/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/TokenExchangeGrantHandler.java

```diff
@@ -0,0 +1,326 @@
+                } catch (NumberFormatException e) {
+                    validityPeriod = TokenExchangeConstants.DEFAULT_IAT_VALIDITY_PERIOD;
+                    log.warn("Invalid value: " + validityPeriodProp + " is set for IAT validity period. Using " +
+                            "default value: " + validityPeriod + " minutes.");
+                }
+            } else {
+                validityPeriod = TokenExchangeConstants.DEFAULT_IAT_VALIDITY_PERIOD;
+                log.warn("Empty value is set for IAT validity period. Using default value: " + validityPeriod
+                        + " minutes.");
+            }
+        } else {
+            log.debug("IAT Validation is disabled for JWT");
+        }
+
+        String registeredClaims = IdentityUtil.getProperty(TokenExchangeConstants.REGISTERED_CLAIMS);
+        if (StringUtils.isNotBlank(registeredClaims)) {
+            registeredClaimNames = registeredClaims.split("\\s*,\\s*");
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug("Validate IAT is set to: " + validateIAT + " for JWT grant.");
+            if (validateIAT) {
+                log.debug("IAT validity period is set to: " + validityPeriod + " minutes for JWT grant.");
+            }
+        }
+    }
+
+    /**
+     * Validate the Token Exchange Grant.
+     * Checks whether the token request satisfies the requirements to exchange the token.
+     *
+     * @param tokReqMsgCtx OAuthTokenReqMessageContext
+     * @return true or false if the grant_type is valid or not.
+     * @throws IdentityOAuth2Exception Error when validating the Token Exchange Grant
+     */
+    @Override
+    public boolean validateGrant(OAuthTokenReqMessageContext tokReqMsgCtx) throws IdentityOAuth2Exception {
+
+        SignedJWT signedJWT;
+        IdentityProvider identityProvider;
+        String tokenEndPointAlias;
+        JWTClaimsSet claimsSet = null;
+        String requested_audience = null;
+        boolean audienceFound;
+
+        RequestParameter[] params = tokReqMsgCtx.getOauth2AccessTokenReqDTO().getRequestParameters();
+        Map<String, String[]> requestParams = Arrays.stream(params).collect(Collectors.toMap(RequestParameter::getKey,
+                RequestParameter::getValue));
+        if (requestParams.get(TokenExchangeConstants.REQUESTED_TOKEN_TYPE) != null) {
+            requested_token_type = requestParams.get(TokenExchangeConstants.REQUESTED_TOKEN_TYPE)[0];
+        }
+        if (requestParams.get(TokenExchangeConstants.AUDIENCE) != null) {
+            requested_audience = requestParams.get(TokenExchangeConstants.AUDIENCE)[0];
+        }
+        String tenantDomain = tokReqMsgCtx.getOauth2AccessTokenReqDTO().getTenantDomain();
+        if (StringUtils.isEmpty(tenantDomain)) {
+            tenantDomain = MultitenantConstants.SUPER_TENANT_DOMAIN_NAME;
+        }
+        if (TokenExchangeConstants.JWT_TOKEN_TYPE.equals(requested_token_type)) {
```

Comment:
Do we need this if check since we are restricting this from the grant validator?

Reply:
Yes, in grant validator we only checked the required parameters ( `subject_token_type`, `subject_token` and ` grant_type`. This is `requested_token_type`, an optional parameter. It is checked here